futures-util/src/stream/futures_unordered/iter_mut.rs

```diff
@@ -1,33 +1,21 @@
-use std::marker::{PhantomData, Unpin};
-use super::FuturesUnordered;
-use super::node::Node;
+use std::marker::Unpin;
+use std::mem::PinMut;
+
+use super::iter_pin_mut::IterPinMut;
 
 #[derive(Debug)]
 /// Mutable iterator over all futures in the unordered set.
-pub struct IterMut<'a, F: 'a + Unpin> {
-    pub(super) node: *const Node<F>,
-    pub(super) len: usize,
-    pub(super) _marker: PhantomData<&'a mut FuturesUnordered<F>>
-}
+pub struct IterMut<'a, F: 'a + Unpin> (pub(super) IterPinMut<'a, F>);
 
 impl<'a, F: Unpin> Iterator for IterMut<'a, F> {
     type Item = &'a mut F;
 
     fn next(&mut self) -> Option<&'a mut F> {
-        if self.node.is_null() {
-            return None;
-        }
-        unsafe {
-            let future = (*(*self.node).future.get()).as_mut().unwrap();
-            let next = *(*self.node).next_all.get();
-            self.node = next;
-            self.len -= 1;
-            Some(future)
-        }
+        PinMut::new(&mut self.0).next().map(|f| unsafe { PinMut::get_mut(f) })
```

Comment:
Why the `PinMut::new` here? You should be able to remove that and have `self.0.next().map(|f| unsafe { PinMut::get_mut(f) })`

Reply:
Yeah ^^'